### PR TITLE
New effect checking system

### PR DESCRIPTION
**Pull request by gilbo**
I'm sure I'm doing this horribly wrong, but hopefully I can find a better process soon.

I don't mean to be requesting review of the commits from the `master` branch that got mixed in here...

### CODE REVIEW

**Comment by yamaguchi1024** [comment]
Is it possible to rebase on top of the commits? This pull request is showing the changes of 11 other previous commits.

**Comment by alexreinking** [comment]
Yeah, this should definitely be rebased before review.

File: SYS_ATL/LoopIR_scheduling.py [hunk 1]
```diff
@@ -2657,6 +2744,9 @@ def map_s(self, s):
         return super().map_s(s)
 
 
+# Note: This analysis is overly conservative.
+# However, it might be a bit involved to come up with
+# a more precise analysis.  TODO
```
Comment (alexreinking) [inline SYS_ATL/LoopIR_scheduling.py:2749]: If you start a comment with `# TODO:` it is easier to automatically track (e.g. with `grep`).

**Comment by alexreinking** [comment]
@gilbo - can you merge `master` into this branch and resolve the conflicts? We should get a move on reviewing this.

File: tests/gemmini/gemmini-rocc-tests/mlps/gemmini_matmul_generator.ipynb [hunk 2]
```diff
@@ -2,7 +2,7 @@
  "cells": [
   {
    "cell_type": "code",
-   "execution_count": 13,
```
Comment (alexreinking) [inline tests/gemmini/gemmini-rocc-tests/mlps/gemmini_matmul_generator.ipynb:5]: Is this change at all relevant?

**Comment by alexreinking** [comment]
The tests need to be updated to use `golden`.

File: tests/test_schedules.py [hunk 3]
```diff
@@ -839,3 +839,21 @@ def foo(n: size, x: R[n]):
 
     foo = foo.lift_if('if n < 20: _')
     assert str(foo) == golden
+                
+
+def test_stage_mem():
```
Comment (alexreinking) [inline tests/test_schedules.py:844]: Should be made into a `golden` test.

File: tests/test_new_eff.py [hunk 4]
```diff
@@ -152,3 +151,141 @@ def foo( N : size, x : R[N,N] ):
 # for its body.  This can probably distinguish whether certain
 # rewrites are allowed or not.
 
+
+def test_delete_config_basic():
```
Comment (alexreinking) [inline tests/test_new_eff.py:155]: Should be made into a `golden` test.

File: tests/test_new_eff.py [hunk 5]
```diff
@@ -152,3 +151,141 @@ def foo( N : size, x : R[N,N] ):
...
+    def foo( N : size, x : R[N] ):
+        CFG.a = 3
+        for i in seq(0,N):
+            x[i] = x[i] + 1.0
+
+    foo = foo.delete_config('CFG.a = _')
+    print(foo)
+
+
+def test_delete_config_subproc_basic():
```
Comment (alexreinking) [inline tests/test_new_eff.py:171]: Should be made into a `golden` test.

File: tests/test_new_eff.py [hunk 6]
```diff
@@ -152,3 +151,141 @@ def foo( N : size, x : R[N,N] ):
...
+            if i < CFG.a:
+                x[i] = x[i] + 1.0
+
+    with pytest.raises(SchedulingError,
+                       match='Cannot change configuration value of CFG_a'):
+        foo = foo.delete_config('do_config()')
+        print(foo)
+
+
+def test_delete_config_bc_shadow():
```
Comment (alexreinking) [inline tests/test_new_eff.py:233]: Should be made into a `golden` test.

File: tests/test_new_eff.py [hunk 7]
```diff
@@ -152,3 +151,141 @@ def foo( N : size, x : R[N,N] ):
...
+        CFG.a = 3
+        for i in seq(0,N):
+            if i < CFG.a:
+                x[i] = x[i] + 1.0
+
+    foo = foo.delete_config('CFG.a = _ #0')
+    print(foo)
+
+
+def test_delete_config_bc_redundant():
```
Comment (alexreinking) [inline tests/test_new_eff.py:251]: Should be made into a `golden` test.

File: tests/test_new_eff.py [hunk 8]
```diff
@@ -152,3 +151,141 @@ def foo( N : size, x : R[N,N] ):
...
+    with pytest.raises(SchedulingError,
+                       match='Cannot change configuration value of CFG_a'):
+        foo = foo.delete_config('CFG.a = _ #1')
+        print(foo)
+
+
+
+
+
+
```
Comment (alexreinking) [inline tests/test_new_eff.py:291]: Don't check in pointless newlines.

File: SYS_ATL/new_eff.py [hunk 9]
```diff
@@ -445,6 +453,7 @@ def globenv_proc(proc):
     'sym':      lambda x: isinstance(x, Sym),
     'aexpr':    lambda x: isinstance(x, A.expr),
     'aenv':     lambda x: isinstance(x, AEnv),
+    'type':     lambda x: type(x) is tuple or T.is_type(x),
```
Comment (alexreinking) [inline SYS_ATL/new_eff.py:456]: `isinstance(x, (tuple, LoopIR.type))`

**Review by alexreinking — CHANGES_REQUESTED** [review]

File: tests/gemmini/gemmini-rocc-tests/mlps/gemmini_matmul_generator.ipynb (hunk 2 above)
Comment (gilbo) [reply tests/gemmini/gemmini-rocc-tests/mlps/gemmini_matmul_generator.ipynb:5]: I don't think so to me unless it was speeding up a test?

File: tests/test_new_eff.py (hunk 8 above)
Comment (gilbo) [reply tests/test_new_eff.py:291]: I do a lot of navigation through files by scrolling, so usually gratuitous newlines are equivalent to a page break of sorts

File: SYS_ATL/new_eff.py (hunk 9 above)
Comment (gilbo) [reply SYS_ATL/new_eff.py:456]: Feel free to change

File: tests/test_new_eff.py (hunk 8 above)
Comment (alexreinking) [reply tests/test_new_eff.py:291]: That's a problem with your editor, then... this doesn't belong in the source.

Comment (alexreinking) [reply tests/test_new_eff.py:291]: IIRC you use Sublime Text? https://stackoverflow.com/a/44185534/2137996

File: SYS_ATL/new_eff.py [hunk 10]
```diff
@@ -1262,6 +1273,290 @@ def bds(x,hi):
...
+        pt_e = ABool(pt.name) if pt.typ == T.bool else AInt(pt.name)
+        is_unchanged    = AImplies(p, G(AEq(sG0(pt_e), sG1(pt_e))))
+        is_read_post    = is_elem(pt, RdGp)
+        is_overwritten  = is_elem(pt, WrGp)
+
+        safe_write  = AImplies( AMay(is_read_post), ADef(is_unchanged) )
+        if not slv.verify(safe_write):
+            slv.pop()
+            raise SchedulingError(
+                f"Cannot rewrite at {stmts[0].srcinfo} because the "
```
Comment (alexreinking) [inline SYS_ATL/new_eff.py:1446]: There is no variable in scope named `stmts`. This should be fixed and a test should be added that covers this code.

File: SYS_ATL/new_eff.py [hunk 11]
```diff
@@ -510,7 +519,8 @@ def _lsstr(ls, prec=0):
         return "∅"
     elif isinstance(ls, LS.Point):
         coords = ','.join([str(a) for a in ls.coords])
-        return f"{{({ls.name},{coords})}}"
+        typ    = f":{ls.type}" if ls.type else ""
+        return f"{{({ls.name},{coords}{typ})}}"
```
Comment (alexreinking) [inline SYS_ATL/new_eff.py:523]: On line 550 below, `return f"alloc({ls.name},{arg})"` references `arg` which does not exist.

File: SYS_ATL/new_eff.py [hunk 12]
```diff
@@ -553,7 +563,7 @@ def is_elem(pt, ls, win_map=None, alloc_masks=None):
     if isinstance(ls, LS.Empty):
         return ABool(False)
     elif isinstance(ls, LS.Point):
-        lspt        = APoint(ls.name, ls.coords)
+        lspt        = APoint(ls.name, ls.coords, ls.type)
```
Comment (alexreinking) [inline SYS_ATL/new_eff.py:566]: On line 578 before `bufname` is misspelled as `bufnane`.

File: SYS_ATL/new_eff.py [hunk 13]
```diff
@@ -1262,6 +1273,290 @@ def bds(x,hi):
...
+
+    # get the set of config variables potentially modified by
+    # the statement block being focused on.  Filter out any
+    # such configuration variables whose values are definitely unchanged
+    def is_cfg_unmod_by_stmts(pt):
+        pt_e    = ABool(pt.name) if pt.typ == T.bool else AInt(pt.name) 
+        #cfg_unwritten = ADef( ANot(is_elem(pt, WrG)) )
+        cfg_unchanged = ADef( G(AEq(pt_e, stmtsG(pt_e))) )
+        return slv.verify(cfg_unchanged)
+    cfg_mod = set( pt.name for pt in get_point_exprs(WrG)
```
Comment (alexreinking) [inline SYS_ATL/new_eff.py:1364]: This variable is never used, but it seems important...

File: SYS_ATL/new_eff.py [hunk 14]
```diff
@@ -1262,6 +1273,290 @@ def bds(x,hi):
...
+                           if not is_cfg_unmod_by_stmts(pt) )
+
+    # consider every global that might be modified
+    cfg_mod_visible = set()
+    for pt in get_point_exprs(WrG):
+        pt_e    = ABool(pt.name) if pt.typ == T.bool else AInt(pt.name) 
+        is_written      = is_elem(pt, WrG)
```
Comment (alexreinking) [inline SYS_ATL/new_eff.py:1371]: This variable is never used, but it seems important...

File: SYS_ATL/new_eff.py [hunk 15]
```diff
@@ -1262,6 +1273,290 @@ def bds(x,hi):
...
+        safe_write  = AImplies( AMay(is_read_post), ADef(is_unchanged) )
+        if not slv.verify(safe_write):
+            # for s in stmts:
+            #     print(s)
+            # print('####')
+            # print(safe_write)
+            # print('***')
+            # print(G)
+            # print('***')
+            # print(stmtsG)
```
Comment (alexreinking) [inline SYS_ATL/new_eff.py:1387]: I'm not going to mark every instance of this... please clean up the commented print debugs.

File: SYS_ATL/new_eff.py [hunk 16]
```diff
@@ -510,7 +519,8 @@ def _lsstr(ls, prec=0):
         return "∅"
     elif isinstance(ls, LS.Point):
         coords = ','.join([str(a) for a in ls.coords])
-        return f"{{({ls.name},{coords})}}"
+        typ    = f":{ls.type}" if ls.type else ""
+        return f"{{({ls.name},{coords}{typ})}}"
     elif isinstance(ls, LS.WholeBuf):
         coords = ','.join([str(a) for a in ls.coords])
         return f"{{{ls.name}|{ls.ndim}}}"
```
Comment (alexreinking) [inline SYS_ATL/new_eff.py:526]: `coords` here is unused, but probably shouldn't be?

File: SYS_ATL/new_eff.py [hunk 17]
```diff
@@ -1262,6 +1273,290 @@ def bds(x,hi):
...
+
+    slv     = SMTSolver(verbose=False)
+    slv.push()
+    slv.assume(AMay(p))
+
+    e0 = lift_e(expr0)
+    e1 = lift_e(expr1)
+
+    test    = G(AEq(e0, e1))
+    is_ok   = slv.verify(test)
```
Comment (alexreinking) [inline SYS_ATL/new_eff.py:1472]: `is_ok` is unused, but probably shouldn't be...

File: SYS_ATL/new_eff.py [hunk 18]
```diff
@@ -202,6 +206,10 @@ def __call__(self, arg):
                     res = ALet(bd.names, bd.rhs, res)
                 else: assert False, "bad case"
             return res
+        elif isinstance(arg, list) and len(arg) == 0:
```
Comment (alexreinking) [inline SYS_ATL/new_eff.py:209]: On line 179 above, `merged = False` is defined, but the variable is never used.

**Review by alexreinking — CHANGES_REQUESTED** [review]

File: tests/test_new_eff.py (hunk 8 above)
Comment (gilbo) [reply tests/test_new_eff.py:291]: This isn't a code editor thing; and I didn't realize it was at the end of the file; it's fine to trim these lines off the end.  Sorry for this going on so long.

File: SYS_ATL/new_analysis_core.py [hunk 19]
```diff
@@ -377,6 +381,112 @@ def pop():
         assert False, "bad case"
 
 
+def aeNegPos(e,pos,env=None,res=None):
```
[the remark this answers is not in the document]
Comment (gilbo) [reply SYS_ATL/new_analysis_core.py:384]: The `push()`, `pop()`, `env` and `res` stuff is a lightweight version of a compiler pass FWIW.  I'm removing `push()` and `pop()` because they are dead code.

I'm sorry this code is hard to follow.  It's trying to implement a reverse dataflow in order to determine whether or not particular code positions are positive or negative.  This is essential for the bug fix we've talked about.

File: SYS_ATL/proc_eqv.py [hunk 20]
```diff
@@ -19,98 +19,154 @@
...
+    assert_eqv_proc(orig_proc, new_proc, config_set)
 
+def assert_eqv_proc(proc1, proc2, config_set=frozenset()):
+    assert isinstance(config_set, frozenset)
+    # First, expand the set of equivalences being tracked if needed
+    for key in config_set:
+        if key not in _UF_Unv_key:
+            new_uf_by_eqv_key(key)
+    # then do the appropriate union operations
+    if len(config_set) == 0:
```
[the remark this answers is not in the document]
Comment (gilbo) [reply SYS_ATL/proc_eqv.py:139]: Until we have some kind of typing to check the argument, I'd prefer to keep this in a form that's more defensive to passing the wrong type for an argument.

File: tests/test_schedules.py [hunk 21]
```diff
@@ -839,3 +839,90 @@ def foo(n: size, x: R[n]):
...
+        for i in seq(0,n/4):
+            for j in seq(0,n/4):
+                for k in seq(0,n/4):
+                    for ii in seq(0,4):
+                        for jj in seq(0,4):
+                            for kk in seq(0,4):
+                                A[4*i+ii,4*j+jj] += ( B[4*i+ii,4*k+kk] *
+                                                      B[4*k+kk,4*j+jj] )
+
+    with pytest.raises(Exception,
```
[the remark this answers is not in the document]
Comment (gilbo) [reply tests/test_schedules.py:879]: Changed.

File: SYS_ATL/new_eff.py [hunk 22]
```diff
@@ -595,44 +605,51 @@ def is_elem(pt, ls, win_map=None, alloc_masks=None):
...
             if ls.name in win_map:
                 lspt    = win_map[ls.name](lspt)
 
             if lspt.name not in alloc_masks:
-                all_bufs[lspt.name] = len(lspt.coords)
+                all_bufs[lspt.name] = (len(lspt.coords),ls.type)
         elif isinstance(ls, LS.WholeBuf):
             bufname     = ls.name
             if bufname in win_map:
                 bufnane = win_map[bufname].name
```
[the remark this answers is not in the document]
Comment (gilbo) [reply SYS_ATL/new_eff.py:627]: I  fixed the typo in upcoming push; sorry for not adding a test right now.

File: SYS_ATL/new_eff.py (hunk 10 above)
Comment (gilbo) [reply SYS_ATL/new_eff.py:1446]: Thanks; did not add test.  Sorry

File: SYS_ATL/new_eff.py (hunk 11 above)
Comment (gilbo) [reply SYS_ATL/new_eff.py:523]: Thanks; just caught this yesterday

File: SYS_ATL/new_eff.py (hunk 13 above)
Comment (gilbo) [reply SYS_ATL/new_eff.py:1364]: yup; copied down to the loop below it.  Thanks

File: SYS_ATL/new_eff.py (hunk 14 above)
Comment (gilbo) [reply SYS_ATL/new_eff.py:1371]: cfg_mod_visible is used in the code I have...?  Oh, this is the point I copied the use of cfg_mod down to.  Thanks.

File: SYS_ATL/new_eff.py (hunk 17 above)
Comment (gilbo) [reply SYS_ATL/new_eff.py:1472]: Thanks

File: SYS_ATL/proc_eqv.py (hunk 20 above)
[the remark this answers is not in the document]
Comment (alexreinking) [reply SYS_ATL/proc_eqv.py:139]: You literally check the type with `isinstance` like three lines up

[the remark this answers is not in the document]
Comment (gilbo) [reply SYS_ATL/proc_eqv.py:139]: Ahh, cool.  I changed it as you suggested now

File: SYS_ATL/proc_eqv.py [hunk 23]
```diff
@@ -19,98 +19,154 @@
...
+            # path splitting optimization
+            grandparent         = self.lookup[parent]
+            self.lookup[val]    = grandparent
+            val, parent         = parent, grandparent
+        return val
+
+    def union(self, val1, val2):
+        p1, p2 = self.find(val1), self.find(val2)
+
+        if p1 is p2:
```
[the remark this answers is not in the document]
Comment (alexreinking) [reply SYS_ATL/proc_eqv.py:96]: That is a good alternative.

**Comment by alexreinking** [comment]
@gilbo - I'm happy with this. Feel free to click the merge button and we can start basing our future work on top of this (and start using the new staging directive!)
